test/integration/health_check_integration_test.cc

```diff
@@ -253,6 +253,46 @@ TEST_P(HttpHealthCheckIntegrationTest, SingleEndpointTimeoutHttp) {
   EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
 }
 
+// Tests that health checking gracefully handles a NO_ERROR GOAWAY from the upstream.
+TEST_P(HttpHealthCheckIntegrationTest, SingleEndpointGoAway) {
+  const uint32_t cluster_idx = 0;
+  initialize();
+  initHttpHealthCheck(cluster_idx);
+
+  // GOAWAY doesn't exist in HTTP1.
+  if (upstream_protocol_ != FakeHttpConnection::Type::HTTP1) {
+    // Send a GOAWAY with NO_ERROR and then a 200. The health checker should allow the request
+    // to finish despite the GOAWAY.
+    clusters_[cluster_idx].host_fake_connection_->onGoAway(Http::GoAwayErrorCode::NoError);
+  }
+  clusters_[cluster_idx].host_stream_->encodeHeaders(
+      Http::TestResponseHeaderMapImpl{{":status", "200"}}, true);
+
+  test_server_->waitForCounterGe("cluster.cluster_1.health_check.success", 1);
+  EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.success")->value());
+  EXPECT_EQ(0, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
+}
+
+// Tests that health checking properly handles a GOAWAY with an error, followed by a reset.
+TEST_P(HttpHealthCheckIntegrationTest, SingleEndpointGoAwayErrorAndReset) {
+  const uint32_t cluster_idx = 0;
+  initialize();
+  initHttpHealthCheck(cluster_idx);
+
+  // GOAWAY doesn't exist in HTTP1.
+  if (upstream_protocol_ != FakeHttpConnection::Type::HTTP1) {
+    // Send a GOAWAY with NO_ERROR and then a 200. The health checker should allow the request
+    // to finish despite the GOAWAY.
+    clusters_[cluster_idx].host_fake_connection_->onGoAway(Http::GoAwayErrorCode::Other);
+  }
+  clusters_[cluster_idx].host_stream_->onResetStream(Http::StreamResetReason::RemoteReset,
```

Comment:
I believe this is the case you were talking about wanting an end to end test for @alyssawilk

Reply:
IMO it would be better to just guard this test and the above test for not HTTP/1, and avoid trying to test multiple things here.

Also, for HTTP/2, I'm not a spec expert, but I'm pretty sure that when sending GOAWAY with error != NOERROR it generally means a connection level error that will be immediately followed by closure, so I'm not sure this test is super accurate.

I think a better integration test for this would be to 1) Just do it for HTTP/2 and 2) Send a GOAWAY != NOERROR and make sure it's counted as a failure and the connection gets recreated?

@alyssawilk WDYT?

Reply:
Yeah, H2 only SGTM.

Reply:
Thanks all, I'll update to only test H2.

@mattklein123 clarifying your last comment, what would "immediately followed by closure" look like? wouldn't that be a remote reset? I do think there's value in making sure a reset after an error goaway isn't double counted as a health check failure just not sure how to make that realistic. I think with the new code we most likely get a LocalReset because we close the connection when we see a GOAWAY with an error, so maybe that would be more realistic?

Reply:
What I meant is that a compliant impl should close the connection after sending GOAWAY != NOERROR. So in your test I would send GOAWAY != NOERROR, the close the connection. This should result in reset of the stream and disconnect, and then you could test that the next health check reconnects?